File: src/main/java/com/xenon/simplyrecipes/views/pages/RecipesView.java
package com.xenon.simplyrecipes.views.pages;

import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.component.html.H2;
import com.vaadin.flow.component.html.Main;
import com.vaadin.flow.component.progressbar.ProgressBar;
import com.vaadin.flow.router.*;
import com.xenon.simplyrecipes.entities.Recipe;
import com.xenon.simplyrecipes.services.CategoryService;
import com.xenon.simplyrecipes.services.RecipeService;
import com.xenon.simplyrecipes.views.MainLayout;
import com.xenon.simplyrecipes.views.components.RecipeCardLayout;
import lombok.SneakyThrows;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.List;

@PageTitle("Recipes")
@Route(value = "recipes", layout = MainLayout.class)
public class RecipesView extends Main implements HasUrlParameter<String> {

    @Autowired
    private RecipeService recipeService;

    @Autowired
    private CategoryService categoryService;

    private String categoryName;

    public RecipesView(RecipeService recipeService, CategoryService categoryService) {
        this.recipeService = recipeService;
        this.categoryService = categoryService;
    }

    @Override
    public void setParameter(BeforeEvent event, @OptionalParameter String categoryName) {
        this.categoryName = categoryName;
        simulateProgressBar();
        initialize();
    }

    @SneakyThrows
    private void simulateProgressBar() {
        ProgressBar progressBar = new ProgressBar();
        add(progressBar);
        progressBar.setIndeterminate(true);
        Thread.sleep(2000);
        progressBar.setIndeterminate(false);
        remove(progressBar);
    }

    private void initialize() {
        addClassName("page-content");

        Div wrapper = new Div();
        wrapper.addClassName("recipe-wrapper");
        getListOfRecipes().forEach(recipe -> wrapper.add(new RecipeCardLayout(recipe)));

        add(getPageTitle(), wrapper);
    }

    private H2 getPageTitle() {
        String titleText = categoryName == null || categoryName.isEmpty()
                ? "Recipes"
                : categoryName + " Recipes";
        H2 title = new H2(titleText);
        title.addClassName("recipe-header");
        return title;
    }

    private List<Recipe> getListOfRecipes() {
        return categoryName == null || categoryName.isEmpty()
                ? recipeService.getAllRecipes()
                : recipeService.getRecipesByCategory(categoryName);
    }

}

File: frontend/themes/my-theme/php/RecipesView.php
<?php
require_once 'vendor/autoload.php'; // Include the necessary autoload file

use com\xenon\simplyrecipes\entities\Recipe;
use com\xenon\simplyrecipes\services\CategoryService;
use com\xenon\simplyrecipes\services\RecipeService;
use com\xenon\simplyrecipes\views\MainLayout;
use com\xenon\simplyrecipes\views\components\RecipeCardLayout;

class RecipesView extends Main implements HasUrlParameter
{
    private $recipeService;
    private $categoryService;
    private $categoryName;

    public function __construct(RecipeService $recipeService, CategoryService $categoryService)
    {
        $this->recipeService = $recipeService;
        $this->categoryService = $categoryService;
    }

    public function setParameter(BeforeEvent $event, $categoryName)
    {
        $this->categoryName = $categoryName;
        $this->initialize();
    }

    private function initialize()
    {
        $this->addClassName("page-content");

        $wrapper = new Div();
        $wrapper->addClassName("recipe-wrapper");
        foreach ($this->getListOfRecipes() as $recipe) {
            $wrapper->add(new RecipeCardLayout($recipe));
        }

        $this->add($this->getPageTitle(), $wrapper);
    }

    private function getPageTitle()
    {
        $titleText = empty($this->categoryName) ? "Recipes" : $this->categoryName . " Recipes";
        $title = new H2($titleText);
        $title->addClassName("recipe-header");
        return $title;
    }

    private function getListOfRecipes()
    {
        return empty($this->categoryName) ? $this->recipeService->getAllRecipes() : $this->recipeService->getRecipesByCategory($this->categoryName);
    }
}
